File: src/libs/cpp/functions/dist.cpp
#include "dist.hpp"

void dist(Context &context) noexcept
{
	auto params = std::get<std::vector<Value>>(context.stack.pop());
	auto point1 = params[0].to_array();
	auto point2 = params[1].to_array();

	// Get the euclidean distance between two points of any dimension.
	// If the points have different dimensions, the lesser dimension is used.
	double total = 0;
	for (size_t i = 0; i < std::min(point1.size(), point2.size()); i++)
	{
		double diff = point1[i].to_number() - point2[i].to_number();
		total += diff * diff;
	}
	context.stack.push(sqrt(total));
}

File: src/runtime/functions/dist.lua
return function(vm)
	local v = vm.pop()
	local b, a = v[1], v[2]
	local t1, t2 = type(a), type(b)
	local result

	if t1 ~= 'table' and t2 == 'table' then
		b = b[1]
	elseif t1 == 'table' and t2 ~= 'table' then
		a = a[1]
	end

	if t1 == 'table' then
		local total = 0
		for i = 1, math.min(#a, #b) do
			local p = a[i] - b[i]
			total = total + p * p
		end
		result = math.sqrt(total)
	else
		result = math.abs(b - a)
	end
	vm.push(result)
end
